scipy/stats/_stats_py.py

```diff
@@ -100,13 +100,20 @@ def _contains_nan(a, nan_policy='propagate'):
         try:
             contains_nan = np.any(np.isnan(a))
         except TypeError:
-            # Don't know what to do. Fall back to omitting nan values and
-            # issue a warning.
-            contains_nan = False
-            nan_policy = 'omit'
-            warnings.warn("The input array could not be properly "
-                          "checked for nan values. nan values "
-                          "will be ignored.", RuntimeWarning)
+            try:
+                # This can happen when attempting to check nan with np.isnan
+                # for string array (e.g. as in the function `rankdata`).
+                a = np.array(a)
```

Comment:
This was caused by gh-16209? Does the logic that was here before gh-16209 work? If so, why not use that? (If this is better, how so?)

Reply:
>This was caused by https://github.com/scipy/scipy/pull/16209, right?

Yes.
> Does the logic that was here before https://github.com/scipy/scipy/pull/16209 still work?

No. 
As #16208, the logic before https://github.com/scipy/scipy/pull/16209 was 
`contains_nan = np.nan in set(a.ravel())`.
But it does not work as expected:
```
In [1]: import numpy as np

In [2]: data = np.array([1,2,3,np.nan])

In [3]: np.nan in set(data.ravel())
Out[3]: False
```
I think the current implementation using `isnan` is OK.
```
In [5]: np.any(np.isnan(data.ravel()))
Out[5]: True
```

 but it does not support a string array. So I added one more try layer.

Reply:
Does the original implementation still work for _string_arrays? If so, why not use it in this third layer?

Reply:
No. The original implementation does not work in both number arrays and string arrays.
```
In [1]: data = np.array([1,2,"3",np.nan])
In [2]: np.nan in set(data.ravel())
Out[2]: False
```

Reply:
I know. 
The next thing I'd ask is whether numerical arrays make it to this third layer. But i can check myself if you'd prefer that I drop it

Reply:
Before https://github.com/scipy/scipy/pull/16209, all tests in gh-16140 passed.
After https://github.com/scipy/scipy/pull/16209, a test in gh-16140 fails.
From that perspective, you could say that "This [failure] was caused by https://github.com/scipy/scipy/pull/16209", "yes".

In that case, "Does the logic that was here before https://github.com/scipy/scipy/pull/16209 still work?" means "can you add an extra `try` block with the old logic:
```python3
               contains_nan = np.nan in set(a.ravel())
```
instead of what you have here
```
                a = np.array(a)
                nan_str = np.array([np.nan]).astype(str)[0]
                contains_nan = a[np.where(a.astype(str) == nan_str)].size != 0
```
and still pass the tests in gh-16140. The answer to that is "yes".

On the other hand, these new tests added in this PR, gh-16238, would fail before and after gh-16140. So from that perspective, "No, this was _not_ caused by https://github.com/scipy/scipy/pull/16209" and "No, the old logic does not work because it never did."

Hopefully, this explains why I was confused!

Now that I understand the state of things, I can assess what to do - use the same logic as before so that all the tests in  gh-16140 would pass, or upgrade the logic so that all these new tests pass.